[backport 7.10] CLDSRV-173 CLDSRV-170 CLDSRV-177 S3C-5390

jonathan-gramain opened the pull request:
Backporting fixes that were dropped accidentally in regular maintenance branches

watsaqat committed:
S3C-5390 s3api head-object with part-number 1 on empty file fails: httpCode 416

(cherry picked from commit 4f3195a6ca6990636bd268b45aa509ae82e137bc)
(cherry picked from commit 8f3e7376647445e0ee2c9493c7991263f810ad8d)

jonathan-gramain committed:
bugfix: CLDSRV-170 skip orphan cleanup in UploadPart[Copy]

Do not delete orphan data in UploadPart/UploadPartCopy on overwrite
iff a CompleteMPU of the target MPU is already in progress.

This is to prevent a race condition where a CompleteMPU is running
while UploadPart is uploading a part for the same MPU.

It leaves an orphan in storage since only one of the upload data will
be present in the finished MPU, but the window is limited to the
CompleteMPU execution and should only occur when there are retries of
UploadPart due to prior stuck requests, or with broken clients
misusing the MPU API, so it should be acceptable.

Implementation details:

- set a flag in the MPU overview key when starting the CompleteMPU
  process, before listing the parts from metadata to construct the
  manifest

- in UploadPart/UploadPartCopy, after the part metadata is written and
  if the same part already existed, re-fetch the MPU overview key to
  check the flag: if set, skip the deletion of the old data of this
  part, since the CompleteMPU process in progress may choose either
  part data depending on the exact timing of the listing vs. the
  part overwrite.

(cherry picked from commit 84961115183cec724c7a9946dea4889f24c734a7)
(cherry picked from commit c2dbbfa0087078b793c0a92cab62dfe68436e7d1)

watsaqat committed:
remove .only

(cherry picked from commit b4725aa03241f296808bdf53f747ec1b1306eb3f)
(cherry picked from commit af95fea311249dd1d81de891265229a246276aa4)

nicolas2bert committed:
CLDSRV-173 DeleteMarkers created by Lifecycle should not be replicated

(cherry picked from commit d30430a81c2e0bbba48a014e907f95272c50d303)

jonathan-gramain committed:
bugfix: CLDSRV-177 fix crash with empty object replication

Fix a case of crash when a replication occurs with an empty object
over a non-empty object.

It is not clear how this happens in practice but there can be some
corner cases with race conditions between object replication and
versioning getting suspended on the target bucket at the same time,
that could lead to this situation, as the check between replication
configuration and actual replication is not atomic.

(cherry picked from commit a4e8cbebe6e7c1be1ac7efdbe1ce6c666a0ad5a2)
(cherry picked from commit 108d1c920f946cb7f4237a5fcb0093345483da4d)

jonathan-gramain committed:
[7.10] CLDSRV-177 add missing test helper checkObjectData

(cherry picked from commit 413ebe743ce3522efe433362a7b19212c1083729)
(cherry picked from commit a1c4420eab4e30e4388c0212bd2d8f82ae28d204)

jonathan-gramain committed:
CLDSRV-173 CLDSRV-170 CLDSRV-177 S3C-5390 bump version

bert-e commented:
# Hello jonathan-gramain,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/cloudserver/doc/user).





*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-173 contains:


* `7.10.15.3`

* `7.10.21`

* `7.10.3.2`

* `7.10.8.5`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `7.10.21`

* `7.70.12`

* `8.4.18`

* `8.5.19`

* `8.6.4`

* `8.7.1`


Please check the `Fix Version/s` of CLDSRV-173, or the target
branch of this pull request.

jonathan-gramain commented:
ping

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/7.70/bugfix/CLDSRV-173-CLDSRV-170-CLDSRV-177-S3C-5390-development-7.10` with contents from `bugfix/CLDSRV-173-CLDSRV-170-CLDSRV-177-S3C-5390-development-7.10`
and `development/7.70`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/7.70/bugfix/CLDSRV-173-CLDSRV-170-CLDSRV-177-S3C-5390-development-7.10 origin/development/7.70
 $ git merge origin/bugfix/CLDSRV-173-CLDSRV-170-CLDSRV-177-S3C-5390-development-7.10
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/7.70/bugfix/CLDSRV-173-CLDSRV-170-CLDSRV-177-S3C-5390-development-7.10
```

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/8.4/bugfix/CLDSRV-173-CLDSRV-170-CLDSRV-177-S3C-5390-development-7.10` with contents from `w/7.70/bugfix/CLDSRV-173-CLDSRV-170-CLDSRV-177-S3C-5390-development-7.10`
and `development/8.4`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/8.4/bugfix/CLDSRV-173-CLDSRV-170-CLDSRV-177-S3C-5390-development-7.10 origin/development/8.4
 $ git merge origin/w/7.70/bugfix/CLDSRV-173-CLDSRV-170-CLDSRV-177-S3C-5390-development-7.10
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.4/bugfix/CLDSRV-173-CLDSRV-170-CLDSRV-177-S3C-5390-development-7.10
```

bert-e commented:
# History mismatch



Merge commit #c1b98f26d01d28fe451cd3594c60e17c9fcd2110 on the integration branch
`w/7.70/bugfix/CLDSRV-173-CLDSRV-170-CLDSRV-177-S3C-5390-development-7.10` is merging a branch which is neither the current
branch `bugfix/CLDSRV-173-CLDSRV-170-CLDSRV-177-S3C-5390-development-7.10` nor the development branch
`development/7.70`.

It is likely due to a rebase of the branch `bugfix/CLDSRV-173-CLDSRV-170-CLDSRV-177-S3C-5390-development-7.10` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

jonathan-gramain commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/7.70/bugfix/CLDSRV-173-CLDSRV-170-CLDSRV-177-S3C-5390-development-7.10` with contents from `bugfix/CLDSRV-173-CLDSRV-170-CLDSRV-177-S3C-5390-development-7.10`
and `development/7.70`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/7.70/bugfix/CLDSRV-173-CLDSRV-170-CLDSRV-177-S3C-5390-development-7.10 origin/development/7.70
 $ git merge origin/bugfix/CLDSRV-173-CLDSRV-170-CLDSRV-177-S3C-5390-development-7.10
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/7.70/bugfix/CLDSRV-173-CLDSRV-170-CLDSRV-177-S3C-5390-development-7.10
```

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/8.4/bugfix/CLDSRV-173-CLDSRV-170-CLDSRV-177-S3C-5390-development-7.10` with contents from `w/7.70/bugfix/CLDSRV-173-CLDSRV-170-CLDSRV-177-S3C-5390-development-7.10`
and `development/8.4`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/8.4/bugfix/CLDSRV-173-CLDSRV-170-CLDSRV-177-S3C-5390-development-7.10 origin/development/8.4
 $ git merge origin/w/7.70/bugfix/CLDSRV-173-CLDSRV-170-CLDSRV-177-S3C-5390-development-7.10
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.4/bugfix/CLDSRV-173-CLDSRV-170-CLDSRV-177-S3C-5390-development-7.10
```

jonathan-gramain committed:
CLDSRV-177 fix backbeat func tests for CLDSRV-290

Since CLDSRV-290, backbeat routes functional tests need to pass a
versionId in the query string in the `/_/backbeat/metadata` route,
hence the original tests from CLDSRV-177 needed to be updated.

bert-e commented:
# History mismatch



Merge commit #98150b2218e644896226aa395cee4f1c2f95aed7 on the integration branch
`w/7.70/bugfix/CLDSRV-173-CLDSRV-170-CLDSRV-177-S3C-5390-development-7.10` is merging a branch which is neither the current
branch `bugfix/CLDSRV-173-CLDSRV-170-CLDSRV-177-S3C-5390-development-7.10` nor the development branch
`development/7.70`.

It is likely due to a rebase of the branch `bugfix/CLDSRV-173-CLDSRV-170-CLDSRV-177-S3C-5390-development-7.10` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

jonathan-gramain commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/7.70/bugfix/CLDSRV-173-CLDSRV-170-CLDSRV-177-S3C-5390-development-7.10` with contents from `bugfix/CLDSRV-173-CLDSRV-170-CLDSRV-177-S3C-5390-development-7.10`
and `development/7.70`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/7.70/bugfix/CLDSRV-173-CLDSRV-170-CLDSRV-177-S3C-5390-development-7.10 origin/development/7.70
 $ git merge origin/bugfix/CLDSRV-173-CLDSRV-170-CLDSRV-177-S3C-5390-development-7.10
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/7.70/bugfix/CLDSRV-173-CLDSRV-170-CLDSRV-177-S3C-5390-development-7.10
```

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/8.4/bugfix/CLDSRV-173-CLDSRV-170-CLDSRV-177-S3C-5390-development-7.10` with contents from `w/7.70/bugfix/CLDSRV-173-CLDSRV-170-CLDSRV-177-S3C-5390-development-7.10`
and `development/8.4`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/8.4/bugfix/CLDSRV-173-CLDSRV-170-CLDSRV-177-S3C-5390-development-7.10 origin/development/8.4
 $ git merge origin/w/7.70/bugfix/CLDSRV-173-CLDSRV-170-CLDSRV-177-S3C-5390-development-7.10
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.4/bugfix/CLDSRV-173-CLDSRV-170-CLDSRV-177-S3C-5390-development-7.10
```

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/CLDSRV-173-CLDSRV-170-CLDSRV-177-S3C-5390-development-7.10` into
`development/7.10`
* [w/7.70/bugfix/CLDSRV-173-CLDSRV-170-CLDSRV-177-S3C-5390-development-7.10](https://github.com/scality/cloudserver/compare/development/7.70...w/7.70/bugfix/CLDSRV-173-CLDSRV-170-CLDSRV-177-S3C-5390-development-7.10) will be merged into `development/7.70` 
* [w/8.4/bugfix/CLDSRV-173-CLDSRV-170-CLDSRV-177-S3C-5390-development-7.10](https://github.com/scality/cloudserver/compare/development/8.4...w/8.4/bugfix/CLDSRV-173-CLDSRV-170-CLDSRV-177-S3C-5390-development-7.10) will be merged into `development/8.4` 
* [w/8.5/bugfix/CLDSRV-173-CLDSRV-170-CLDSRV-177-S3C-5390-development-7.10](https://github.com/scality/cloudserver/compare/development/8.5...w/8.5/bugfix/CLDSRV-173-CLDSRV-170-CLDSRV-177-S3C-5390-development-7.10) will be merged into `development/8.5` 
* [w/8.6/bugfix/CLDSRV-173-CLDSRV-170-CLDSRV-177-S3C-5390-development-7.10](https://github.com/scality/cloudserver/compare/development/8.6...w/8.6/bugfix/CLDSRV-173-CLDSRV-170-CLDSRV-177-S3C-5390-development-7.10) will be merged into `development/8.6` 
* [w/8.7/bugfix/CLDSRV-173-CLDSRV-170-CLDSRV-177-S3C-5390-development-7.10](https://github.com/scality/cloudserver/compare/development/8.7...w/8.7/bugfix/CLDSRV-173-CLDSRV-170-CLDSRV-177-S3C-5390-development-7.10) will be merged into `development/8.7` 



The following branches will **NOT** be impacted:

* `development/7.4`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

jonathan-gramain commented:
/approve

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/cloudserver/commit/7be27e0a83dc644a9ec232daf8df09f22f7a2bb4) did not succeed in branch w/8.4/bugfix/CLDSRV-173-CLDSRV-170-CLDSRV-177-S3C-5390-development-7.10.




*The following options are set:* **approve**

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/cloudserver/commit/461f5ac5f9bebd6db3b33095e79aa720ad8c125e) did not succeed in branch w/8.5/bugfix/CLDSRV-173-CLDSRV-170-CLDSRV-177-S3C-5390-development-7.10.




*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/7.10`

* :heavy_check_mark: `development/7.70`

* :heavy_check_mark: `development/8.4`

* :heavy_check_mark: `development/8.5`

* :heavy_check_mark: `development/8.6`

* :heavy_check_mark: `development/8.7`



The following branches will **NOT** be impacted:

* `development/7.4`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/7.10`

* :heavy_check_mark: `development/7.70`

* :heavy_check_mark: `development/8.4`

* :heavy_check_mark: `development/8.5`

* :heavy_check_mark: `development/8.6`

* :heavy_check_mark: `development/8.7`



The following branches have **NOT** changed:

* `development/7.4`



Please check the status of the associated issue CLDSRV-173.

Goodbye jonathan-gramain.
